First version of overview generator

mihails-strasuns-sociomantic opened the pull request:
This is essentially port of old overview generator tool from backend repository updated to use GitHub API (via octod) for fetching all information.

This, however, has brought some complications because some information I have maintained manually before is not currently deducible in automated manner (and I can't put it in public repo). I plan to introduce YAML dependency to neptune and use it to define basic project metadata format (to include d2-ready and is-library flags initially).

As such, don't expect tool in this PR to actually work now :)

mihails-strasuns-sociomantic commented:
It hardly comes as surprise that only D YAML library (https://github.com/kiith-sa/D-YAML) doesn't compile with latest DMD versions. Let's talk about this on Monday.

mihails-strasuns-sociomantic commented:
We can proceed with review/merge of this part while dependency issues is being resolved.

leandro-lucarella-sociomantic commented on src/overview/htmlreport.d:
```diff
@@ -0,0 +1,147 @@
+/*******************************************************************************
+
```
A module header would be nice.

leandro-lucarella-sociomantic commented on src/overview/main.d:
```diff
@@ -1,8 +1,91 @@
+/*******************************************************************************
+
```
Same.

leandro-lucarella-sociomantic commented on src/overview/main.d:
```diff
@@ -1,8 +1,91 @@
+
+    Configuration conf;
+    conf.dryRun = false;
+    conf.oauthToken = "READ-ME-FROM-CONFIG";
```
I don't know exactly how this is used but it would be nice for this tool to be able to run without authentication (you can still access public repos as a unauthenticated user).

leandro-lucarella-sociomantic commented on src/overview/main.d:
```diff
@@ -1,8 +1,91 @@
+    import std.exception : ifThrown;
+
+    auto repos = client
+        .listOrganizationRepos("sociomantic")
```
I think this should be moved to a command-line argument / config file. It makes no sense to publish a script that has an internal / private organization hardcoded (same for exclude projects, etc.)

leandro-lucarella-sociomantic commented on src/overview/repository.d:
```diff
@@ -0,0 +1,145 @@
+    /// current versions of dependencies (key is dependency name)
+    Version[string] dependencies;
+    /// is project tested with D2 in Jenkins?
+    bool            d2ready;
```
Mmm, I see this is still too sociomantic specific. I had the idea this versioning definition / tooling was going to be generic. I think we might think a bit more how to do specific and general stuff separately. Maybe make this d2ready thing optional via config / cmd-line options too, so at least we keep it as a hack for us and for now.

leandro-lucarella-sociomantic commented:
@mathias-lang-sociomantic I can't ask for a review from you using GH facilities for some reason, can you have a look? :)

mathias-lang-sociomantic commented on src/overview/htmlreport.d:
```diff
@@ -0,0 +1,147 @@
+    dependency version string put into intersection cell.
+
+    Params:
+        project = mapping of project names to matching repo metadata
```
Mismatch between parameter name and doc

mathias-lang-sociomantic commented on src/overview/repository.d:
```diff
@@ -0,0 +1,145 @@
+/*******************************************************************************
+
+    Utilities to work fetch single repository metadata
```
Not a very clear description :)

leandro-lucarella-sociomantic commented:
> Now it depends on which level we want it to be generic: for example, dependencies handling will probably be Makd specific. But the library / D2 ready part as well as org and excluded projects don't feel so nice.

Yes, it is to be expected to follow some conventions to be able to use the tools, like using Makd, or submodules or having a YML file (although we should define the expected fields here). That's all fine.

mihails-strasuns-sociomantic commented:
Updated:
- Plug in sociomantic-tsunami/D-YAML
- Check `.metadata.yml` to see if project is a library
- Create simplistic configuration reader and move sociomantic specific bits there

mihails-strasuns-sociomantic commented:
(Have yet to clean up coding style and address earlier comments)

leandro-lucarella-sociomantic commented on Build.mak:
```diff
@@ -19,7 +19,7 @@ $O/libdyaml.a: $(shell find $C/submodules/dyaml/source -type f -name '*.d') \
 ################################################################################
 
-$O/%unittests: override LDFLAGS += -L$O -lvibed -levent -lssl -lcrypto
+$O/%unittests: override LDFLAGS += -L$O -lvibed -ldyaml -levent -lssl -lcrypto
```
Maybe squash this into the commit that introduced it.

leandro-lucarella-sociomantic commented:
I don't know why so many comments are marked as outdated, it was done against the latest commits :unamused:

leandro-lucarella-sociomantic commented on src/overview/main.d:
```diff
@@ -1,8 +1,141 @@
+        .assocArray();
+
+    octod_conf.dryRun = entries["dry_run"].to!bool.ifThrown(true);
+    octod_conf.oauthToken = entries["token"];
```
BTW, I would also name this `oauthtoken` to make it a bit more clear.

mihails-strasuns-sociomantic commented on .gitmodules:
```diff
@@ -13,3 +13,9 @@
 [submodule "openssl"]
 	path = submodules/openssl
 	url = https://github.com/D-Programming-Deimos/openssl.git
+[submodule "dyaml"]
```
no reason, changed to `d-yaml`

mihails-strasuns-sociomantic commented on .gitmodules:
```diff
@@ -13,3 +13,9 @@
+[submodule "dyaml"]
+	path = submodules/dyaml
+	url = https://github.com/sociomantic-tsunami/d-yaml
+[submodule "tunyendian"]
```
dependency of d-yaml (fixed name to `tinyendian`)

mihails-strasuns-sociomantic commented on src/overview/repository.d:
```diff
@@ -120,36 +125,50 @@ void updateRepositoryDependencies ( ref Repository project, HTTPConnection clien
-    alias submodules = repo => repo.download("submodules/").expectDirectory.listAll();
+    try
+    {
+        alias submodules = repo => repo.download("submodules/").expectDirectory.listAll();
```
I don't have git repo so can't use `git` command. Parsing `.gitmodules` shouldn't be too much of a problem though.

leandro-lucarella-sociomantic replied:
You don't need a git repo to use `git config -f <file>`. But I guess the git config ini format is simple enough to parse, is just that I prefer to rely on tools in case there is any changes in formats.

mihails-strasuns-sociomantic commented:
Updated:

- fixed all style/docs issues
- switch to YAML for own config
- use .gitmodules
- remove d2ready bits

New config file looks like this:

```yaml
dry_run: false
oauthtoken: XXXXX

organization: sociomantic
include:
    - sociomantic-tsunami/makd
    - sociomantic-tsunami/ocean
exclude:
    - tango
    - phobos
    - chest
    - druntime
    - dtools
```

This has grown from a quick script into quite an app by now :)

leandro-lucarella-sociomantic commented on src/overview/submodules.d:
```diff
@@ -0,0 +1,132 @@
+    import std.typecons : Yes;
+    import std.string : strip;
+
+    static rgxSection = regex(`^\[submodule "(.+)"\]$`);
```
I guess there is no *ini* parser in phobos...

mihails-strasuns-sociomantic commented:
Anything that I have missed?
